core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,182 @@
+  MaybeReferenceCountedPtr(T* value_ptr, Deleter deleter)
+      : m_value_ptr(value_ptr), m_control(new Control{std::move(deleter), 1}) {}
+
+ public:
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      MaybeReferenceCountedPtr&& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+    other.m_value_ptr = nullptr;
+    other.m_control   = nullptr;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      const MaybeReferenceCountedPtr& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      MaybeReferenceCountedPtr&& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr       = other.m_value_ptr;
+      other.m_value_ptr = nullptr;
+      m_control         = other.m_control;
+      other.m_control   = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      const MaybeReferenceCountedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr = other.m_value_ptr;
+      m_control   = other.m_control;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (is_reference_counted())
+        Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~MaybeReferenceCountedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_value_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept {
+    KOKKOS_EXPECTS(bool(*this));
+    return *get();
+  }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the MaybeReferenceCountedPtr manages the lifetime of the
+  // object
+  KOKKOS_FUNCTION bool is_reference_counted() const noexcept {
+    return m_control != nullptr;
+  }
+
+ private:
+  KOKKOS_FUNCTION void cleanup() noexcept {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    // If m_counter is set, then this instance is responsible for managing the
+    // objects pointed to by m_counter and m_value_ptr.
+    if (is_reference_counted()) {
+      KOKKOS_EXPECTS(m_control->m_deleter);
+      int const count = Kokkos::atomic_fetch_sub(&(m_control->m_counter), 1);
+      if (count == 1) {
+        (m_control->m_deleter)(m_value_ptr);
+        m_value_ptr = nullptr;
+        delete m_control;
+        m_control = nullptr;
+      }
+    }
+#endif
+  }
+
+  struct Control {
+    std::function<void(T*)> m_deleter;
+    int m_counter;
+  };
+
+  T* m_value_ptr;
+
+ protected:
+  Control* m_control;
+};
+
+template <class T>
+class HostSharedPtr : public MaybeReferenceCountedPtr<T> {
+ public:
+  explicit HostSharedPtr(T* value_ptr = nullptr)
```

Comment:
I'm with @dalg24 on this one; we shouldn't be reference-counting a null pointer value.

Why not just add a regular default constructor (to both `HostSharedPtr` and a `protected` one to `MaybeReferenceCountedPtr`), and a `KOKKOS_EXPECTS` in this one?

We can even add a deleted constructor for `std::nullopt_t`.

(I generally find separate constructors easier than default parameters, even if their implementations just forward to more general constructors.)

Reply:
Hmm... IMHO, it's more useful for `HostSharedPtr` and `UnmanagedPtr` to be semi-regular than to disallow assigning `nullptr` (in particular since `std::shared_ptr` also allows this).